src/event/value.rs

```diff
@@ -312,6 +342,964 @@ impl Value {
+    ///
+    /// let val = Value::from(1);
+    /// assert_eq!(val.is_empty(), false);
+    ///
+    /// let mut val = Value::from(Vec::<Value>::default());
+    /// assert_eq!(val.is_empty(), true);
+    /// val.insert(0, 1);
+    /// assert_eq!(val.is_empty(), false);
+    /// val.insert(3, 1);
+    /// assert_eq!(val.is_empty(), false);
+    ///
+    /// let mut val = Value::from(BTreeMap::default());
+    /// assert_eq!(val.is_empty(), true);
+    /// val.insert("foo", 1);
+    /// assert_eq!(val.is_empty(), false);
+    /// val.insert("bar", 2);
+    /// assert_eq!(val.is_empty(), false);
+    /// ```
+    #[instrument(level = "trace")]
+    pub fn is_empty(&self) -> bool {
+        match &self {
+            Value::Boolean(_)
+            | Value::Bytes(_)
+            | Value::Timestamp(_)
+            | Value::Float(_)
+            | Value::Integer(_) => false,
+            Value::Null => true,
+            Value::Map(v) => v.is_empty(),
+            Value::Array(v) => v.is_empty(),
+        }
+    }
+
+    /// Return the number of subvalues the value has.
+    ///
+    /// ```rust
+    /// use shared::{event::*, lookup::*};
+    /// use std::collections::BTreeMap;
+    ///
+    /// let val = Value::from(1);
+    /// assert_eq!(val.len(), None);
+    ///
+    /// let mut val = Value::from(Vec::<Value>::default());
+    /// assert_eq!(val.len(), Some(0));
+    /// val.insert(0, 1);
+    /// assert_eq!(val.len(), Some(1));
+    /// val.insert(3, 1);
+    /// assert_eq!(val.len(), Some(4));
+    ///
+    /// let mut val = Value::from(BTreeMap::default());
+    /// assert_eq!(val.len(), Some(0));
+    /// val.insert("foo", 1);
+    /// assert_eq!(val.len(), Some(1));
+    /// val.insert("bar", 2);
+    /// assert_eq!(val.len(), Some(2));
+    /// ```
+    #[instrument(level = "trace")]
+    pub fn len(&self) -> Option<usize> {
+        match &self {
+            Value::Boolean(_)
+            | Value::Bytes(_)
+            | Value::Timestamp(_)
+            | Value::Float(_)
+            | Value::Integer(_)
+            | Value::Null => None,
+            Value::Map(v) => Some(v.len()),
+            Value::Array(v) => Some(v.len()),
+        }
+    }
+
+    fn insert_coalesce(
+        sub_segments: Vec<FieldBuf>,
+        working_lookup: LookupBuf,
+        sub_value: &mut Value,
+        value: Value,
+    ) -> std::result::Result<Option<Value>, EventError> {
+        // Creating a needle with a back out of the loop is very important.
+        let mut needle = None;
+        for sub_segment in sub_segments {
+            let mut lookup = LookupBuf::from(sub_segment);
+            lookup.extend(working_lookup.clone()); // We need to include the rest of the insert.
+                                                   // Notice we cannot take multiple mutable borrows in a loop, so we must pay the
+                                                   // contains cost extra. It's super unfortunate, hopefully future work can solve this.
```

Comment:
Could the clone be avoided with a `Lookup<'a>`? BTW you should be able to take multiple immutable borrows.

Reply:
I don't think it's any better:

```rust     
        for sub_segment in sub_segments {
            let segment: Segment = sub_segment.name.as_str().into();
            let mut lookup = Lookup::from(segment);
            let new_working_lookup: Lookup = (&working_lookup).into();
            lookup.extend(new_working_lookup); // We need to include the rest of the insert.

            if !sub_value.contains(lookup) {
                let mut lookup: LookupBuf = LookupBuf::from(sub_segment);
                lookup.extend(working_lookup);
                needle = Some(lookup);
                break;
            }
        }
```
        
There's still a number of allocations, it might be slightly better than the clone, but that's questionable. I'm sure it could be improved further and it's probably worth keeping in mind for future work, but for now I think we should leave it as it is..

Reply:
Any thoughts on (another) lookup reference type to help avoid some clones, possibly for future work? This might allow for working better with `AsRef` or `Deref`.
```rust
enum LookupRef<'a> {
    Lookup(Lookup<'a>), // (or maybe needs to be &'a Lookup<'a> ?)
    LookupBuf(&'a LookupBuf),
}
```
Just thinking aloud, mostly.